Add information about implied dependencies and detached plugin (#5958)
jenkins-infra/jenkins.io: content/doc/book/managing/plugins.adoc

on startup.

== What is an implied dependency?

Features are sometimes detached (or split off) from Jenkins core and moved into a plugin.

Many plugins, like plugin:subversion[Subversion] and plugin:junit[JUnit], got their beginnings as Jenkins core functionalities.
When a plugin was attached to Jenkins core prior to being detached, it may or may not have used its full functionality with other plugins that relied on the same version of Jenkins.
To ensure that plugins do not fail when a functionality on which they rely is separated from Jenkins core, it is necessary to have a dependency in the detached plugin, if it specifies a dependency on a version of Jenkins core prior to the split.
It is assumed that there is a dependency on the detached plugin, even when it isn't stated explicitly.

Implied dependencies build up over time for plugins that don't frequently update which Jenkins core version they depend on.

=== Why does the implied dependency exist?

It is "implied" since it is not stated explicitly in the plugin, and it is a dependency because Jenkins core is uncertain if the APIs available in Jenkins core, prior to the separation of the plugin, are necessary or not.

> For example, because the WMI Windows Agent plugin was separated from Jenkins core, Jenkins core is uninformed if the dependent plugin needs any functionality that was previously present in Jenkins core.

Reviewer: identity instance instead of WMI?

Author: Yes my bad
i'll correct it

This creates an implied dependency.

As a Jenkins administrator, you can see the plugins that have an implied dependency from the plugin manager page. 
Hover over the "uninstall" button and a list of plugins with an implied dependency is displayed.

image::managing/hover-for-implied-dependencies.png["Hover over the uninstall button to see Implied Dependencies", role=center]

=== How can the implied dependency be removed?

> The implied dependency can be removed by releasing a new version of the plugin that depends on a newer minimum Jenkins core than Jenkins 1.547. 

Reviewer: ```suggestion
The implied dependency can be removed by releasing a new version of the plugin that depends on a newer minimum Jenkins core.
```

present, Jenkins will use whatever plugin version the user has specified.
If the file is absent, Jenkins will restore the plugin to the default version
on startup.

== What is an implied dependency?

Features are sometimes detached (or split off) from Jenkins core and moved into a plugin.

Many plugins, like plugin:subversion[Subversion] and plugin:junit[JUnit], got their beginnings as Jenkins core functionalities.

> When a plugin was attached to the Jenkins core before being detached, it may or may not have used any of its functionality in other plugins that relied on that version of Jenkins.
> To guarantee that plugins do not fail when the functionality on which they rely is separated from Jenkins core, it is regarded to have a dependency on the detached plugin if it specifies a dependency on a version of Jenkins core prior to the split.

Reviewer 2: ```suggestion
When a plugin was attached to Jenkins core prior to being detached, it may or may not have used its full functionality with other plugins that relied on the same version of Jenkins.
To ensure that plugins do not fail when a functionality on which they rely is separated from Jenkins core, it is necessary to have a dependency in the detached plugin, if it specifies a dependency on a version of Jenkins core prior to the split.
```
update of syntax/content to try and make the text flow better, if this takes away from the message let me know and we can go from there!

Author: looks good

> It is assumed that there is a dependency on the detached plugin even though it isn't stated explicitly.
> 
> Implied dependencies build up over time for plugins that don't frequently update which Jenkins core version they depend on.

Reviewer 2: ```suggestion
It is assumed that there is a dependency on the detached plugin, even when it isn't stated explicitly.

Implied dependencies build up over time for plugins that don't frequently update which Jenkins core version they depend on.
```
small syntax suggestion

> === Why does the implied dependency exist?
> 
> Because the WMI Windows Agent plugin was separated from Jenkins core, Jenkins core is uninformed if the dependent plugin needs any functionality that was previously present in Jenkins core.
> This creates an implied dependency.
> 
> It is "implied" since it is not stated explicitly in the plugin, and it is a dependency because Jenkins core is uncertain if the APIs available in Jenkins core prior to the separation of the plugin are necessary or not.

Reviewer 2: ```suggestion
=== Why does the implied dependency exist?

It is "implied" since it is not stated explicitly in the plugin, and it is a dependency because Jenkins core is uncertain if the APIs available in Jenkins core, prior to the separation of the plugin, are necessary or not.

For example, because the WMI Windows Agent plugin was separated from Jenkins core, Jenkins core is uninformed if the dependent plugin needs any functionality that was previously present in Jenkins core.
This creates an implied dependency.

```
rearranged content to provide the logic behind implied dependencies first, and then adding a qualifier of "For example" to the example used since Windows WMI is not mentioned prior to this on the page.

[NOTE]
====

> Release 1.547 of Jenkins separated the WMI Windows Agent plugin from the main code.

Reviewer 2: ```suggestion
The release 1.547 of Jenkins separated the WMI Windows Agent plugin from the main code.
```
small syntax suggestion

Jenkins core assumes a dependency on plugins that made a declaration of dependence on a previous version of Jenkins core.
====

> The Jenkins administrator can see the plugins that have an implied dependency from the plugin manager page. 
> They hover over the "uninstall" button and a list of plugins with an implied dependency is displayed.

Reviewer 2: ```suggestion
As a Jenkins administrator, you can see the plugins that have an implied dependency from the plugin manager page. 
Hover over the "uninstall" button and a list of plugins with an implied dependency is displayed.
```
This would be a good candidate to have a screenshot attached. Displaying what the uninstall hover shows would help immensely for users to know for sure where they should be looking.

Author: i'll add some screenshots then

=== How can the implied dependency be removed?

The implied dependency can be removed by releasing a new version of the plugin that depends on a newer minimum Jenkins core than Jenkins 1.547. 

> The link:https://www.jenkins.io/doc/developer/tutorial-improve/[Improve a plugin tutorial] provides steps that can assist plugin developers as they configure the plugin to link:https://www.jenkins.io/doc/developer/tutorial-improve/update-base-jenkins-version/[depend on a newer minimum Jenkins core].

Reviewer 2: ```suggestion
The link:/doc/developer/tutorial-improve/[Improve a plugin tutorial] provides steps that can assist plugin developers as they update the plugin to link:/doc/developer/tutorial-improve/update-base-jenkins-version/[depend on a newer minimum Jenkins core].
```
Syntax suggestion and update to links so that they are utilizing existing Jenkins links. This sort of format, `link:/doc/book/installing/[]` can be used for any link that goes to another page in jenkins.io.

> === Which plugins have an implied dependency on WMI Windows Agents plugin?
> 
> As of 16 Dec 2022, there were over 300 Jenkins plugins that declare a minimum Jenkins version before 1.547.
> Those plugins are listed in the link:https://docs.google.com/spreadsheets/d/1KDIJIu41rIRlj6PjzyhAY9vazKmd3PVwqM3_-32Lt-o/edit?usp=sharing["Plugins with implied dependency on WMI Windows Agents" worksheet].
> 
> link:https://issues.jenkins.io/secure/Dashboard.jspa[Jira issues] for those plugins to upgrade their minimum Jenkins version can be linked in this epic as a convenient way of seeing the related issues.
> A Jira issue is not required in order to release a new version of an affected plugin.
> The Jira issue is just a convenience in case it helps someone justify that new release or explain why it is not being released.

Reviewer 2: ```suggestion
=== Which plugins have an implied dependency on WMI Windows Agents plugin?

As of December 16 2022, there were over 300 Jenkins plugins that declare a minimum Jenkins version before 1.547.
Those plugins are listed in the link:https://docs.google.com/spreadsheets/d/1KDIJIu41rIRlj6PjzyhAY9vazKmd3PVwqM3_-32Lt-o/edit?usp=sharing["Plugins with implied dependency on WMI Windows Agents" worksheet].

GitHub issues for plugins to upgrade their minimum Jenkins version can be linked in link:https://issues.jenkins.io/browse/JENKINS-70301[this epic] as a convenient way of seeing the related issues.
A Jira issue is not required in order to release a new version of an affected plugin.
The Jira epic provides a unified location for tracking and can help someone justify the new release or explain why it is not being released.
```
updated the syntax and content here, there were a couple of updates needed for the proper linking/connecting to the epic, and the plugins are more likely than not tracking issues in GitHub, as all the PRs created in the google sheet lead to the respective plugin repository. The epic provides a single place to link to these PRs for convenience.

present, Jenkins will use whatever plugin version the user has specified.
If the file is absent, Jenkins will restore the plugin to the default version
on startup.

== What is an implied dependency?

Features are sometimes detached (or split off) from Jenkins core and moved into a plugin.

Many plugins, like plugin:subversion[Subversion] and plugin:junit[JUnit], got their beginnings as Jenkins core functionalities.
When a plugin was attached to Jenkins core prior to being detached, it may or may not have used its full functionality with other plugins that relied on the same version of Jenkins.

> To ensure that plugins do not fail when a functionality on which they rely is separated from Jenkins core, it is necessary to have a dependency in the detached plugin, if it specifies a dependency on a version of Jenkins core prior to the split.

Reviewer: ```suggestion
To ensure that plugins do not fail when a functionality on which they rely is separated from Jenkins core, plugins that depend on certain version of core are treated as if they were dependent on all the plugins that were parts of that core version and became detached later.
```

Reviewer: this is copy-pasted from plugin-site, I think it's more clear.
